File: scripts/custom-metadata-upload/custom_metadata_upload.rb
#!/usr/bin/env ruby
require 'rubygems'
require 'json'
require 'logger'
require 'optparse'
require 'httparty'
require 'csv'
require 'uri-handler'
require 'pry'

logger = Logger::new("./custom-metadata-upload.log")

options = {}
OptionParser.new do |opts|
  opts.banner = "Usage: custom_field.rb [options]"

  opts.on("-f n", "--file", "CSV Import File") do |f|
    options[:file] = f
  end

  opts.on("-d n", "--debug", "Enables HTTP debugger") do |d|
    options[:debug] = d
  end 

  opts.on("-i n", "--integration-key", "X-Pendo-Integration-Key") do |i|
    options[:integration_key] = i
  end

  opts.on("-t n", "--metadata-type", "'visitor' or 'account'") do |t|
    options[:type] = t
  end
end.parse!

# Read File from Path
file_contents = CSV.parse(File.read(options[:file]), headers: true)
integration_key = options[:integration_key] 
headers = file_contents.headers.drop(1)
request_body_array = []

file_contents.each do |row|
  id = row[0].to_s
  request_entry = {
    "#{options[:type]}Id" => id,
    "values" => {}
    }

  headers.each_with_index do |header, index|
    request_entry["values"][header] = row[index+1] || ""
  end

  request_body_array.push(request_entry)
end
    
endpoint_url = "https://app.pendo.io/api/v1/metadata/#{options[:type]}/custom/value"

http_headers = { 
  "Content-Type" => "application/json",
  "X-Pendo-Integration-Key" => integration_key,
}

begin
  request = HTTParty.post(
    endpoint_url,
    headers: http_headers, 
    body: request_body_array.to_json,
    logger: logger, log_level: :info, log_format: :curl
  )

rescue HTTParty::Error
  logger.error("HTTParty error for #{row[1]}")
rescue StandardError
  logger.error("Standard error for #{row[1]}")
end

File: scripts/customMetadataUpload_JavaScript/custom-metadata-upload.js
var logger = require('logger').createLogger('./custom-metadata-upload.log');
var { Command } = require('commander');
var axios = require('axios');
var Papa = require('papaparse');
var fs = require('fs-extra');
const program = new Command();
let options = {};

program
	.description('Send or update custom metadata for visitors/accounts.')
	.option('-f, --file <string>', 'CSV Import File', (f) => (options.file = f))
	.option('-d', '--debug', 'Enables HTTP debugger', (d) => (options.debug = d))
	.option(
		'-i, --integration-key <string>',
		'X-Pendo-Integration-Key',
		(i) => (options.integration_key = i)
	)
	.option('-t, --metadata-type <string>', "'visitor' or 'account'", (t) => (options.type = t));

program.parse();

// Read File from Path
let csvFile = Papa.parse(fs.readFileSync(options.file, 'utf8'), {
	headers: false,
	worker: false,
});
let file_contents = csvFile.data;
let headers = file_contents.shift().slice(1, 4);
let request_body_array = [];

file_contents.forEach((row) => {
	let id = row[0].toString();
	let request_entry = { [`${options.type}Id`]: id, values: {} };

	headers.forEach((header, index) => (request_entry.values[header] = row[index + 1] || ''));

	request_body_array.push(request_entry);
	return (request_body = JSON.stringify(request_body_array));
});

let apiUrl = `https://app.pendo.io/api/v1/metadata/${options.type}/custom/value`;

let http_headers = {
	'Content-Type': 'application/json',
	'X-Pendo-Integration-Key': options.integration_key,
};

let uploadCustomMetadata = () => {
	axios
		.post(apiUrl, request_body, {
			headers: http_headers,
		})
		.then(function (response) {
			logger.info(response);
		})
		.catch(function (error) {
			logger.error(error);
			console.log(error);
		});
};

uploadCustomMetadata();
